internal/core/application/market_loader_service.go

```diff
@@ -62,21 +63,92 @@ func (m *marketsLoaderService) FetchMarkets() {
 		return
 	}
 
+	//assume that external service, from which we are fetching markets, should return active markets
+	activeMarkets := make(map[string]domain.Market)
 	for _, v := range liquidityProviders {
-		for _, v1 := range v.Markets {
-			if err := m.marketRepository.InsertMarket(context.Background(), domain.Market{
+		for _, market := range v.Markets {
+			mkt := domain.Market{
 				ProviderName: v.Name,
 				Url:          v.Endpoint,
-				BaseAsset:    v1.BaseAsset,
-				QuoteAsset:   v1.QuoteAsset,
-			}); err != nil {
-				log.Errorf("FetchMarkets -> InsertMarket: %v", err)
+				BaseAsset:    market.BaseAsset,
+				QuoteAsset:   market.QuoteAsset,
+				Active:       true,
+			}
+			activeMarkets[mkt.Key()] = mkt
+		}
+	}
+
+	if err := m.updateMarketActiveStatusAndInsertNew(existingMarkets, activeMarkets); err != nil {
+		log.Errorf("FetchMarkets -> updateMarketActiveStatusAndInsertNew: %v", err)
+	}
+}
+
+func (m *marketsLoaderService) updateMarketActiveStatusAndInsertNew(
+	existingMarkets []domain.Market,
+	activeMarkets map[string]domain.Market,
+) error {
+	//since markets can go on and off, we need to update markets accordingly
+	//3 cases here:
+	//1. existing market is not in active markets -> set market as inactive
+	//2. existing market is in active markets -> set market as active
+	//3. active market is not in existing markets -> create new market
+	marketsNotInActiveList := make([]domain.Market, 0) //to be inactivated
+	marketsInActiveList := make([]domain.Market, 0)    //to be activated
+	for _, v := range existingMarkets {
+		if _, ok := activeMarkets[v.Key()]; ok {
+			marketsInActiveList = append(marketsInActiveList, v)
+		} else {
+			marketsNotInActiveList = append(marketsNotInActiveList, v)
+		}
+	}
+
+	if len(marketsNotInActiveList) > 0 {
+		for _, v := range marketsNotInActiveList {
+			if err := m.marketRepository.InactivateMarket(
+				context.Background(),
+				v.ID,
+			); err != nil {
+				return fmt.Errorf("updateMarketActiveStatusAndInsertNew -> InactivateMarket: %v", err)
 			}
 		}
 	}
+
+	if len(marketsInActiveList) > 0 {
+		for _, v := range marketsInActiveList {
+			if err := m.marketRepository.ActivateMarket(
+				context.Background(),
+				v.ID,
+			); err != nil {
+				return fmt.Errorf("updateMarketActiveStatusAndInsertNew -> ActivateMarket: %v", err)
+			}
+		}
+	}
+
+	//now we need to add new markets to db, if market already exist, insert wont do anything
+	for _, v := range activeMarkets {
+		if err := m.marketRepository.InsertMarket(context.Background(), domain.Market{
+			ProviderName: v.ProviderName,
+			Url:          v.Url,
+			BaseAsset:    v.BaseAsset,
+			QuoteAsset:   v.QuoteAsset,
+			Active:       true,
+		}); err != nil {
+			return fmt.Errorf("updateMarketActiveStatusAndInsertNew -> InsertMarket: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func isSameMarket(m1, m2 domain.Market) bool {
+	return m1.Url == m2.Url &&
+		m1.BaseAsset == m2.BaseAsset &&
+		m1.QuoteAsset == m2.QuoteAsset
```

Comment:
this function seems to be unused, can we remove it?

Reply:
Done